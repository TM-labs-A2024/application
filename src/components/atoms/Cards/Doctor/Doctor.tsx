import { Link } from '@chakra-ui/next-js'
import { Text } from '@chakra-ui/react'
import { specialities } from '@constants/index'
import { Doctor as DoctorType } from '@src/types'
import NextLink from 'next/link'
import React from 'react'

import Chevron from '../../../../../public/static/icons/chevron.svg'

export default function Doctor({ doctor }: { doctor: DoctorType }) {
  return (
    <Link as={NextLink} href={`/medico/${doctor.id}`} className="w-full">
      <div className="flex flex-row items-center justify-between">
        <div className="flex flex-col items-start gap-1 overflow-hidden">
          <Text>
            {doctor.firstname} {doctor.lastname}
          </Text>
          <div className="flex flex-row flex-wrap items-center gap-2">
            <Text className="text-nowrap">CI: {doctor.id.toLocaleString('es-ES')}</Text>
            {doctor.specialities?.map((speciality, idx) => (
              <Text key={`doctor-card-${speciality}`} className="text-nowrap">
                {specialities.find((el) => el.id === speciality)?.name}
                {doctor.specialities.length > 0 && idx !== doctor.specialities.length - 1
                  ? ','
                  : ''}
              </Text>
            ))}
          </div>
        </div>
        <Chevron />
      </div>
    </Link>
  )
}
