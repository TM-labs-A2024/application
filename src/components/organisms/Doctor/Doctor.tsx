import { ArrowBackIcon } from '@chakra-ui/icons'
import {
  IconButton,
  Text,
  Stack,
  Heading,
  Divider,
  Button,
  Modal,
  ModalOverlay,
  ModalContent,
  ModalFooter,
  ModalBody,
  useDisclosure
} from '@chakra-ui/react'
import { ACCESS_DENIED, ACCESS_GRANTED, ACCESS_REMOVED } from '@constants/index'
import { specialities } from '@constants/index'
import { Doctor as DoctorType } from '@src/types'
import { isIOS, isMobile } from '@utils/index'
import { format } from 'date-fns'
import { es } from 'date-fns/locale/es'
import { useRouter } from 'next/navigation'
import React, { ReactElement, useCallback } from 'react'
import { Store } from 'react-notifications-component'

function ConfirmationModal({
  isOpen,
  onClose,
  onSubmit,
  method,
  name
}: {
  isOpen: boolean
  onClose: () => void
  onSubmit: () => void
  method: string
  name: string
}) {
  return (
    <Modal isOpen={isOpen} onClose={onClose}>
      <ModalOverlay />
      <ModalContent>
        <ModalBody>
          ¿Seguro que desea {method.toLowerCase()} el acceso a {name}?
        </ModalBody>

        <ModalFooter>
          <Button mr={3} onClick={onClose}>
            Cancelar
          </Button>
          <Button colorScheme="blackAlpha" onClick={onSubmit}>
            {method} acceso
          </Button>
        </ModalFooter>
      </ModalContent>
    </Modal>
  )
}

export default function Doctor({ doctor }: { doctor: DoctorType }): ReactElement {
  // --- Hooks -----------------------------------------------------------------
  const router = useRouter()
  const { isOpen, onOpen, onClose } = useDisclosure()
  const {
    isOpen: isApprovalOpen,
    onOpen: onApprovalOpen,
    onClose: onApprovalClose
  } = useDisclosure()
  const { isOpen: isRemovalOpen, onOpen: onRemovalOpen, onClose: onRemovalClose } = useDisclosure()
  // --- END: Hooks ------------------------------------------------------------

  // --- Local state -----------------------------------------------------------
  // --- END: Local state ------------------------------------------------------

  // --- Refs ------------------------------------------------------------------
  // --- END: Refs -------------------------------------------------------------

  // --- Redux -----------------------------------------------------------------
  // --- END: Redux ------------------------------------------------------------

  // --- Side effects ----------------------------------------------------------
  // --- END: Side effects -----------------------------------------------------

  // --- Data and handlers -----------------------------------------------------
  const onApproval = useCallback(() => {
    Store.addNotification(ACCESS_GRANTED(isMobile(window)))
    onApprovalClose()
    router.push('/medicos')
  }, [onApprovalClose, router])

  const onDenial = useCallback(() => {
    Store.addNotification(ACCESS_DENIED(isMobile(window)))
    onClose()
    router.push('/medicos')
  }, [onClose, router])

  const onRemoval = useCallback(() => {
    Store.addNotification(ACCESS_REMOVED(isMobile(window)))
    onRemovalClose()
    router.push('/medicos')
  }, [onRemovalClose, router])
  // --- END: Data and handlers ------------------------------------------------

  return (
    <div className={`flex h-screen w-screen flex-col p-8 lg:px-96 ${isIOS() ? 'pt-20' : 'pt-8'}`}>
      <div className="mb-8 flex flex-row items-center justify-start gap-4">
        <IconButton
          size="xl"
          aria-label="back"
          variant="link"
          icon={<ArrowBackIcon />}
          onClick={() => {
            router.push('/medicos')
          }}
        />
        <Text className="font-medium">Perfil del médico</Text>
      </div>
      <div className="flex h-full flex-col justify-between">
        <div>
          <Stack spacing={1} mb={6}>
            <Heading as="h3" size="md" noOfLines={1}>
              {doctor.firstname} {doctor.lastname}
            </Heading>
            <Text>CI: {doctor.id.toLocaleString('es-ES')}</Text>
          </Stack>
          <Divider orientation="horizontal" />
          <Stack mb={6} mt={6}>
            <h4 className="text-sm text-gray-600">Especialidad</h4>
            <div className="flex flex-row flex-wrap gap-2">
              {doctor.specialities?.map((speciality, idx) => (
                <Text className="text-nowrap font-medium" key={`doctor-card-${speciality}`}>
                  {specialities.find((el) => el.id === speciality)?.name}
                  {doctor.specialities.length > 0 && idx !== doctor.specialities.length - 1
                    ? ','
                    : ''}
                </Text>
              ))}
            </div>
          </Stack>
          <Stack mb={6}>
            <h4 className="text-sm text-gray-600">Fecha de nacimiento</h4>
            <Text className="font-medium">
              {format(new Date(doctor.birthdate), "dd 'de' MMMM, yyyy", {
                locale: es
              })}
            </Text>
          </Stack>
          <Stack mb={6}>
            <h4 className="text-sm text-gray-600">Correo electrónico</h4>
            <Text className="font-medium">{doctor.email}</Text>
          </Stack>
          <Stack mb={6}>
            <h4 className="text-sm text-gray-600">Teléfono</h4>
            <Text className="font-medium">{doctor.phone}</Text>
          </Stack>
        </div>
        {!doctor?.patientPending && <Button onClick={onRemovalOpen}>Revocar acceso</Button>}
        {doctor?.patientPending && (
          <div className="flex w-full flex-row gap-4">
            <Button className="flex-grow" variant="outline" onClick={onOpen}>
              Rechazar
            </Button>
            <Button className="flex-grow" onClick={onApprovalOpen}>
              Aceptar
            </Button>
          </div>
        )}
      </div>
      <ConfirmationModal
        isOpen={isOpen}
        onClose={onClose}
        onSubmit={onDenial}
        method="Denegar"
        name={`${doctor.firstname} ${doctor.lastname}`}
      />
      <ConfirmationModal
        isOpen={isApprovalOpen}
        onClose={onApprovalClose}
        onSubmit={onApproval}
        method="Conceder"
        name={`${doctor.firstname} ${doctor.lastname}`}
      />
      <ConfirmationModal
        isOpen={isRemovalOpen}
        onClose={onRemovalClose}
        onSubmit={onRemoval}
        method="Revocar"
        name={`${doctor.firstname} ${doctor.lastname}`}
      />
    </div>
  )
}
