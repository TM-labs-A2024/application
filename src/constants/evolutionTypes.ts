export const evolutionTypes = [
  {
    id: '1',
    name: 'Evolución'
  },
  {
    id: '2',
    name: 'Hospitalización'
  },
  {
    id: '3',
    name: 'Alta'
  },
  {
    id: '4',
    name: 'Primera consulta'
  }
]
