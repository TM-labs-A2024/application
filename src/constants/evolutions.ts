export const evolution = {
  specialty: {
    description: '',
    id: '',
    name: ''
  },
  author: '',
  title: '',
  description: '',
  'content-format': '',
  created_at: '',
  payload: {
    type: {
      value: '',
      label: ''
    },
    reason: '',
    description: '',
    history: '',
    examination: '',
    summary: '',
    diagnostic: '',
    comments: ''
  }
}
