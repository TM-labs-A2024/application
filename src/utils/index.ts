import { formatDate } from './dates'
import { sendEmail, generateVerificationCode, setupEmailSending } from './email'
import { setupErrorNotification } from './notifications'
import { isWeb, isIOS, isAndroid, isMobile } from './platform'

export {
  isWeb,
  isIOS,
  isAndroid,
  isMobile,
  sendEmail,
  generateVerificationCode,
  setupEmailSending,
  setupErrorNotification,
  formatDate
}
