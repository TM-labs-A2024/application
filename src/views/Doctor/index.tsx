import Doctor from '@components/organisms/Doctor'
import { Doctor as DoctorType, ReactSelectOption } from '@src/types'
import React from 'react'

export default function DoctorView({
  context
}: {
  context: {
    specialties: ReactSelectOption[]
    doctor: DoctorType
    isDenialOpen: boolean
    onDenialOpen: () => void
    onDenialClose: () => void
    isApprovalOpen: boolean
    onApprovalOpen: () => void
    onApprovalClose: () => void
    isRemovalOpen: boolean
    onRemovalOpen: () => void
    onRemovalClose: () => void
    onApproval: () => void
    onDenial: () => void
    onRemoval: () => void
    isLoading: boolean
  }
}) {
  return <Doctor context={context} />
}
