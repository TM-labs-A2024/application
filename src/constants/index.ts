import { doctors } from './doctors'
import { evolutionTypes } from './evolutionTypes'
import { institution } from './institution'
import { institutions } from './institutions'
import {
  ACCESS_REMOVED,
  FILTERS_APPLIED,
  ACCESS_REQUESTED,
  ATTACHMENT_DELETED,
  IMAGE_DELETED,
  ACCESS_DENIED,
  ACCESS_GRANTED,
  ACCESS_REMOVAL,
  GENERIC_NOTIFICATION,
  GENERIC_ERROR
} from './notifications'
import { nurse } from './nurses'
import { patient, sexOptions, statusOptions } from './patient'
import { patients } from './patients'
import { specialties, specialities } from './specialties'
import { specialtyData } from './specialtyData'

export {
  specialties,
  specialities,
  doctors,
  nurse,
  patient,
  sexOptions,
  statusOptions,
  specialtyData,
  evolutionTypes,
  patients,
  institution,
  institutions,
  ACCESS_REMOVED,
  FILTERS_APPLIED,
  ACCESS_REQUESTED,
  ATTACHMENT_DELETED,
  IMAGE_DELETED,
  ACCESS_DENIED,
  ACCESS_GRANTED,
  ACCESS_REMOVAL,
  GENERIC_NOTIFICATION,
  GENERIC_ERROR
}
