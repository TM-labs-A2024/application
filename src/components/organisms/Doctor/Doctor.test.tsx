import { doctors, specialties } from '@src/constants'
import { formatDate } from '@src/utils'
import { render, screen } from '@test/utils/index'
import { format } from 'date-fns'
import { es } from 'date-fns/locale/es'

import '@testing-library/jest-dom'
import Doctor from './Doctor'
import 'next/navigation'

jest.mock('next/navigation')

const doctor = doctors[0]

let isDenialOpen = false
let isApprovalOpen = false
let isRemovalOpen = false

const onDenialOpen = () => {
  isDenialOpen = true
}

const onApprovalOpen = () => {
  isApprovalOpen = true
}

const onRemovalOpen = () => {
  isRemovalOpen = true
}

const onDenialClose = () => {
  isDenialOpen = false
}

const onApprovalClose = () => {
  isApprovalOpen = false
}

const onRemovalClose = () => {
  isRemovalOpen = false
}

const specialtiesOptions = specialties.map((option: { name: string; id: string }) => ({
  value: option.id,
  label: option.name
}))

const context = {
  specialties: specialtiesOptions,
  doctor,
  isDenialOpen,
  onDenialOpen,
  onDenialClose,
  isApprovalOpen,
  onApprovalOpen,
  onApprovalClose,
  isRemovalOpen,
  onRemovalOpen,
  onRemovalClose,
  onApproval: jest.fn(),
  onDenial: jest.fn(),
  onRemoval: jest.fn(),
  isLoading: false
}

describe('Organisms > Doctor test', () => {
  test('The component shows the name, govId, specialties, birthdate, email and phoneNumberNumber of the doctor', () => {
    render(<Doctor context={context} />)

    const idMatcher = (content: string) => content.includes(doctor.id)
    const doctorId = screen.getByText(idMatcher)

    const emailMatcher = (content: string) => content.includes(doctor.email)
    const doctorEmail = screen.getByText(emailMatcher)

    expect(screen.getByText(`${doctor.firstname} ${doctor.lastname}`)).toBeInTheDocument()
    expect(doctorId).toBeInTheDocument()
    doctor.specialties?.forEach((specialty) => {
      const specialtyMatcher = (content: string) =>
        content.includes(String(specialtiesOptions?.find((el) => el.value === specialty.id)?.label))

      expect(screen.getByText(specialtyMatcher)).toBeInTheDocument()
    })
    expect(
      screen.getByText(
        format(new Date(formatDate(doctor.birthdate)), "dd 'de' MMMM, yyyy", {
          locale: es
        })
      )
    ).toBeInTheDocument()
    expect(doctorEmail).toBeInTheDocument()
    expect(screen.getByText(doctor.phoneNumber)).toBeInTheDocument()
  })

  // test('When clicking on the access removal button, the respective confirmation modal is shown', () => {
  //   const doctorNew = doctors[10]
  //   render(<Doctor context={{ ...context, doctor: doctorNew }} />)

  //   const deletionButton = screen.getByText('Revocar acceso')
  //   fireEvent.click(deletionButton)

  //   const confirmation = `¿Seguro que desea revocar el acceso a ${doctorNew.firstname} ${doctorNew.lastname}?`

  //   expect(screen.getByText(confirmation)).toBeInTheDocument()
  // })

  // test('The component allows to remove access to a doctor', () => {
  //   const doctorNew = doctors[10]
  //   render(<Doctor context={{ ...context, doctor: doctorNew }} />)

  //   const deletionButton = screen.getByText('Revocar acceso')
  //   fireEvent.click(deletionButton)

  //   const confirmation = `¿Seguro que desea revocar el acceso a ${doctorNew.firstname} ${doctorNew.lastname}?`

  //   expect(screen.getByText(confirmation)).toBeInTheDocument()
  //   expect(screen.getAllByText('Revocar acceso')).toHaveLength(2)
  // })

  // test("If the doctor doesn't have access, when clicking on the access approval button, the respective confirmation modal is shown", () => {
  //   render(<Doctor context={context} />)

  //   const deletionButton = screen.getByText('Aceptar')
  //   fireEvent.click(deletionButton)

  //   const confirmation = `¿Seguro que desea conceder el acceso a ${doctor.firstname} ${doctor.lastname}?`

  //   expect(screen.getByText(confirmation)).toBeInTheDocument()
  // })

  // test('The component allows to give access to a doctor', () => {
  //   render(<Doctor context={context} />)

  //   const deletionButton = screen.getByText('Aceptar')
  //   fireEvent.click(deletionButton)

  //   const confirmation = `¿Seguro que desea conceder el acceso a ${doctor.firstname} ${doctor.lastname}?`

  //   expect(screen.getByText(confirmation)).toBeInTheDocument()
  //   expect(screen.getByText('Conceder acceso')).toBeInTheDocument()
  // })

  // test("If the doctor doesn't have access, when clicking on the access denial button, the respective confirmation modal is shown", () => {
  //   render(<Doctor context={context} />)

  //   const deletionButton = screen.getByText('Rechazar')
  //   fireEvent.click(deletionButton)

  //   const confirmation = `¿Seguro que desea denegar el acceso a ${doctor.firstname} ${doctor.lastname}?`

  //   expect(screen.getByText(confirmation)).toBeInTheDocument()
  // })

  // test('The component allows to deny access to a doctor', () => {
  //   render(<Doctor context={context} />)

  //   const deletionButton = screen.getByText('Rechazar')
  //   fireEvent.click(deletionButton)

  //   const confirmation = `¿Seguro que desea denegar el acceso a ${doctor.firstname} ${doctor.lastname}?`

  //   expect(screen.getByText(confirmation)).toBeInTheDocument()
  //   expect(screen.getByText('Denegar acceso')).toBeInTheDocument()
  // })

  test('Matches the snapshot', () => {
    render(<Doctor context={context} />)

    expect(screen.getByTestId('doctor')).toMatchSnapshot()
  })
})
