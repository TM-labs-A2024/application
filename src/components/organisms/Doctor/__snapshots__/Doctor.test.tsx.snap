// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Organisms > Doctor test Matches the snapshot 1`] = `
<div
  class="flex h-screen w-screen flex-col p-8 lg:px-96 pt-8"
  data-testid="doctor"
>
  <div
    class="mb-8 flex flex-row items-center justify-start"
  >
    <button
      aria-label="back"
      class="chakra-button mr-4 css-84vx7q"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
          fill="currentColor"
        />
      </svg>
    </button>
    <p
      class="chakra-text font-medium css-0"
    >
      Perfil del médico
    </p>
  </div>
  <div
    class="flex h-full flex-col justify-between"
  >
    <div>
      <div
        class="chakra-stack css-1uriqkk"
      >
        <h3
          class="chakra-heading css-ypohwm"
        >
          José
           
          Pérez
        </h3>
        <p
          class="chakra-text css-0"
        >
          CI: 
          1000000
        </p>
      </div>
      <hr
        aria-orientation="horizontal"
        class="chakra-divider css-svjswr"
      />
      <div
        class="chakra-stack css-1420a3n"
      >
        <h4
          class="text-sm text-gray-600"
        >
          Especialidad
        </h4>
        <p
          class="chakra-text text-wrap css-0"
        >
          undefined.
        </p>
      </div>
      <div
        class="chakra-stack css-1j35777"
      >
        <h4
          class="text-sm text-gray-600"
        >
          Fecha de nacimiento
        </h4>
        <p
          class="chakra-text font-medium css-0"
        >
          11 de noviembre, 1999
        </p>
      </div>
      <div
        class="chakra-stack css-1j35777"
      >
        <h4
          class="text-sm text-gray-600"
        >
          Correo electrónico
        </h4>
        <p
          class="chakra-text font-medium css-0"
        >
          joseperez@gmail.com
        </p>
      </div>
      <div
        class="chakra-stack css-1j35777"
      >
        <h4
          class="text-sm text-gray-600"
        >
          Teléfono
        </h4>
        <p
          class="chakra-text font-medium css-0"
        >
          4127775522
        </p>
      </div>
    </div>
    <div
      class="flex w-full flex-row"
    >
      <button
        class="chakra-button mr-2 flex-grow css-11494p6"
        type="button"
      >
        Rechazar
      </button>
      <button
        class="chakra-button ml-2 flex-grow css-ez23ye"
        type="button"
      >
        Aceptar
      </button>
    </div>
  </div>
</div>
`
